exporter/server.go

```diff
@@ -97,6 +104,68 @@ func multiTargetHandler(serverMap ServerMap) http.HandlerFunc {
 	}
 }
 
+// OverallTargetsHandler is a handler to scrape all the targets in one request.
+// Adds instance label to each metric.
+func OverallTargetsHandler(exporters []*Exporter, logger *logrus.Logger) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		seconds, err := strconv.Atoi(r.Header.Get("X-Prometheus-Scrape-Timeout-Seconds"))
+		// To support also older ones vmagents.
+		if err != nil {
+			seconds = 10
+		}
+
+		var gatherers prometheus.Gatherers
+		gatherers = append(gatherers, prometheus.DefaultGatherer)
+
+		filters := r.URL.Query()["collect[]"]
+
+		for _, e := range exporters {
+			ctx, cancel := context.WithTimeout(r.Context(), time.Duration(seconds-e.opts.TimeoutOffset)*time.Second)
+			defer cancel()
```

Comment:
These defer (and the one on L135) won't run when each loop ends though, is that intentional?

Reply:
As I understand the logic we create gatherers in the loop, but real metrics collecting is being done by promhttp.HandlerFor which calls collector.Collect and it is located outside the loop, so we free the resources after the all work done.